Migrates report filters

dianabarsan opened the pull request:
# Description

Migrates and refactors report filters. 
Some other scattered fixes and todos. 

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented:
hey @latin-panda , would you mind having a look? Thanks for your time!

latin-panda commented on webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:
```diff
@@ -0,0 +1,189 @@
+import { select, Store } from '@ngrx/store';
+import { Selectors } from '../../../selectors';
+import { combineLatest, Subscription, from } from 'rxjs';
+import { GlobalActions } from '../../../actions/global';
```
```suggestion
import { GlobalActions } from '@mm-actions/global';
```

latin-panda commented on webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:
```diff
@@ -0,0 +1,189 @@
+import { combineLatest, Subscription, from } from 'rxjs';
+import { GlobalActions } from '../../../actions/global';
+import { MultiDropdownFilterComponent } from '@mm-components/filters/multi-dropdown-filter/mullti-dropdown-filter.component';
+import { PlaceHierarchyService } from '../../../services/place-hierarchy.service';
```
```suggestion
import { PlaceHierarchyService } from '@mm-services/place-hierarchy.service';
```

latin-panda reviewed (commented):
This PR sparks joy ✨✨ I like the filters!!
I left few tiny suggestions. But in general everything looks good!
Would you like to include unit test or in the next PR?

dianabarsan commented on webapp/src/ts/components/filters/date-filter/date-filter.component.ts:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  applyFilter() {
+    this.globalActions.setFilter({ date: this.date });
```
You know what, I'm not sure why the filters are global. If it were up to me, I'd have had filters be dedicated. 
Maybe when this was added, there was something that needed access to filters in general? Maybe there still is? 

I guess we'll find out at the end of the migration :)

dianabarsan commented on webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  trackByFn(idx, facility) {
+    return facility.doc._id;
```
I changed it so it null checks the doc, thanks. 
We're not listening to changes here, this list doesn't "react" to changes, so we won't be redrawing items. But I can add the rev as well.

dianabarsan commented on webapp/src/ts/components/filters/facility-filter/facility-filter.component.ts:
```diff
@@ -0,0 +1,189 @@
+}
+
+
+/*angular.module('inboxDirectives').directive('mmFacilityFilter', function(SearchFilters) {
```
It's not complete, there is one todo left which is a performance issue. 
I think it's too early to test the performance of any part of the application, so I'm leaving it as a todo for now and keeping this comment in so I remember.

dianabarsan commented on webapp/src/ts/components/filters/form-type-filter/form-type-filter.component.ts:
```diff
@@ -0,0 +1,65 @@
+  dropdownFilter: MultiDropdownFilterComponent;
+
+  constructor(
+    private store:Store,
```
Is this the stylistic approach? 

I find it really hard to differentiate between Object assigns and type declarations visually, so I keep the space for object property assignments `{ a: 'something' }` and no space when doing type declarations `{ a:string }`.

dianabarsan commented on webapp/src/ts/modules/reports/reports-filters.component.ts:
```diff
@@ -0,0 +1,52 @@
+  @ViewChild(FacilityFilterComponent)
+  facilityFilter:FacilityFilterComponent;
+  @ViewChild(DateFilterComponent)
+  dateFilter:DateFilterComponent;
```
They are generic (and were pretty generic before too). 
We reuse the freetext filter in contacts.

dianabarsan commented on webapp/src/ts/modules/reports/reports.component.ts:
```diff
@@ -191,10 +209,30 @@ export class ReportsComponent implements OnInit, OnDestroy {
+      return;
+    }
+    this.globalActions.clearFilters();
+    // todo reset filters!
```
Thanks for pointing this out! Nope, not needed anymore.

dianabarsan commented:
hey @latin-panda 

I've added/migrated unit tests and changed some minor things as a result. :D 
I also had an attempt to fixing the unit test build job, and it was one thing to fix after another :)) 
But I fixed it, the stupid build works! https://travis-ci.org/github/medic/cht-core/jobs/731598931

anyway, please have another look. Thanks for your time!

latin-panda commented on webapp/src/ts/modules/reports/reports.component.ts:
```diff
@@ -191,7 +209,17 @@ export class ReportsComponent implements OnInit, OnDestroy {
+    // clears report selection for any text search or filter selection
+    // does not clear selection when someone is editing a form
+    if((this.filters.search || Object.keys(this.filters).length > 1) && !this.enketoEdited) {
+      //$state.go('reports.detail', { id: null }, { notify: false });
```
Do we still need this "if"?

dianabarsan replied:
Yea, I think so. It's still commented because we don't have a report detail page yet. :D

latin-panda reviewed (approved):
Excellent very neat!! I don't really have additional comments, just added 1 question inline and I have this other question:

Are the following files going to be migrated/finalised in future PRs? Just wanted to double check that these weren't accidentally forgotten
/tests/karma/ts/pipes/date.spec.js 
/tests/karma/ts/providers/lineage.spec.js  
/tests/karma/ts/reducers/reports.spec.ts

dianabarsan commented:
> Are the following files going to be migrated/finalised in future PRs? Just wanted to double check that these weren't accidentally forgotten
/tests/karma/ts/pipes/date.spec.js
/tests/karma/ts/providers/lineage.spec.js
/tests/karma/ts/reducers/reports.spec.ts

Not forgotten, but I wanted to get this PR over with. I should have added those in the initial branch but tests weren't setup yet :( 
My plan is to: get the detail page working and then migrate all the skipped unit tests (there are a bunch!!)
